Sources/SecureXPC/XPCServerEndpoint.swift

```diff
@@ -0,0 +1,23 @@
+//
+//  XPCServerEndpoint.swift
+//  
+//
+//  Created by Alexander Momchilov on 2021-11-28.
+//
+
+import Foundation
+
+// TODO: make this codable so it can be sent over XPC.
+public struct XPCServerEndpoint {
+    // Technically, an `xpc_endpoint_t` is sufficient to create a new connection, on its own. However, it's useful to
+    // be able to communicate the kind of connection, and its name, so we also store those, separately.
+    internal let kind: XPCConnectionType
+    internal let serviceName: String?
```

Comment:
In what circumstances is it valid for the name to be `nil`?

Reply:
iff it's an anonymous service

Reply:
I think we should structure this as part of the `XPCConnectionType` enum with an associated value such that when a connection must have a connection name, it has one, and when it can't then it's not possible to specify one. That'll let the compiler ensure we've implemented the invariant correctly. Thoughts?

Reply:
I’ll play around with it and let you know how it goes!

Reply:
Oh I remember why I didn’t do this:

1. it complicates the structure of the way this value gets encoded (IIRC Swift added the ability to auth-synthesize the codable conformance for an enumeration with associated values, but I haven’t played with that yet)
2. It would make it impossible to build an XPCEndpoint out of a plain `xpc_endpoint_t`.

The latter comes down to a matter of philosophy. My opinion is that we should allow people to access endpoints, connections, audit tokens, etc, if they want to. I don’t think a framework called “SecureXPC” should be afraid of exposing the fact that is uses XPC under the hood (it’s not a secret ;) ). NSXPCConnection was designed as a more opaque wrapper where everything should “just work”, but as you’ve probably seen, that adds a bit too much smoke and mirrors sometimes.

Reply:
In regards to #1, if all of the associated values are `Codable` then conformance is automatically generated by the compiler. [`XPCError`](https://github.com/trilemma-dev/SecureXPC/blob/main/Sources/SecureXPC/XPCError.swift) make use of this; several of its cases have `String` associated values.

In regards to #2, I consider it an intentional design choice of SecureXPC that it does not expose the C API it uses. I don't have any issue with `NSXPCConnection` in regards to its abstractions except that it's an Objective-C API which maps very awkwardly to Swift. My view is escape hatches cause many issues and should only be done as a last resort. Fundamentally I don't view SecureXPC as a wrapper of the XPC C API, I view it as its own XPC API which internally makes use of the XPC C API. (This is quite different than the [Blessed framework](https://github.com/trilemma-dev/Blessed) I open sourced which is intentionally a Swift-friendly wrapper of the Authorization Services and Service Management frameworks.) Happy to discuss this in more depth in the Discussions topic.